controllers/certificatesigningrequests_controller.go

```diff
@@ -19,45 +19,195 @@ package controllers
+	if err != nil {
+		if k8sapierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, r.reconcileCSR(certificateSigningRequest, configMap)
+}
 
-	// your logic here
+// NewCertificateSigningRequestsReconciler returns a pointer to certificateSigningRequestsReconciler
+func NewCertificateSigningRequestsReconciler(mgr manager.Manager, clusterConfig cluster.Config, watchNamespace string) (*certificateSigningRequestsReconciler, error) {
+	clientset, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating kubernetes clientset")
+	}
 
-	return ctrl.Result{}, nil
+	return &certificateSigningRequestsReconciler{
+			instanceReconciler: instanceReconciler{
+				client:             mgr.GetClient(),
+				log:                ctrl.Log.WithName("controllers").WithName("CertificateSigningRequests"),
+				k8sclientset:       clientset,
+				clusterServiceCIDR: clusterConfig.Network().GetServiceCIDR(),
+				vxlanPort:          "",
+				watchNamespace:     watchNamespace,
+			},
+			platform: clusterConfig.Platform(),
+		},
+		nil
+}
+
+// reconcileCSR handles the CSR validation and approval
+func (r *certificateSigningRequestsReconciler) reconcileCSR(request *certificatesv1.CertificateSigningRequest, configMap *core.ConfigMap) error {
+	if request == nil || configMap == nil {
+		return errors.New("CSR or instance ConfigMap cannot be nil")
+	}
+
+	hostNames, err := r.findHostNames(configMap)
+	if err != nil {
+		return errors.Wrapf(err, "unable to find host names from ConfigMap %s", configMap.Name)
+	}
+
+	certificateSigningRequest, err := csr.NewApprover(request, r.k8sclientset, hostNames, r.log)
+	if err != nil {
+		return errors.Wrapf(err, "could not create WMCO CSR Approver")
+	}
+
+	if err := certificateSigningRequest.Approve(); err != nil {
+		return errors.Wrapf(err, "WMCO CSR Approver could not approve %s CSR", request.Name)
+	}
+
+	return nil
+}
+
+// findHostNames returns a list of the instance host names from the instance configMap
+// depending on the cloud provider type of the cluster.
+func (r *certificateSigningRequestsReconciler) findHostNames(configMap *core.ConfigMap) ([]string, error) {
+	// Get the list of hosts that are expected to be Nodes
+	hosts, err := parseHosts(configMap.Data)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse hosts from ConfigMap %s", configMap.Name)
+	}
+	var requireVMHostName bool
+	// platform type set None or vSphere indicates hostName required is the actual hostName of the VM
+	if r.platform == oconfig.NonePlatformType || r.platform == oconfig.VSpherePlatformType {
+		requireVMHostName = true
+	}
```

Comment:
If the platform type is set to vSphere or None, the future node name that gets constructed is using the actual hostname of the VM, otherwise it is constructed using the instance host address(mapped from instance configmap). I ll try to add a more clear comment, if that's the issue here.

Reply:
So a reverse look up is not needed? I thought from the Slack thread, the direction was to try doing the reverse lookup and if that fails, use the hostname. Or did I misunderstand?

Reply:
Also talked about this, I think it would be great to have the high level CSR approval process documented. I am having a hard time wrapping my head around all the changes made here, and having all the considerations laid out either at the top of the csr.go file, or as a separate doc in the `docs` folder would be very helpful